File: libs/power-log-parser/src/lib/parsers/wheel-of-death-counter-parser.ts
import { CardIds, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, NodeType, ShowEntity, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class WheelOfDeathCounterParser implements ActionParser {
	readonly ParserName = 'WheelOfDeathCounterParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		let tagChange: TagChange | null = null;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(tagChange = node.Object as TagChange).Name === (GameTag.TAG_SCRIPT_DATA_NUM_1 as number) &&
			this.GameState.CurrentEntities.get(tagChange.Entity)?.CardId ===
				CardIds.WheelOfDeath_WheelOfDeathCounterEnchantment_TOY_529e1
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.ShowEntity &&
			(node.Object as ShowEntity).CardId === CardIds.WheelOfDeath_WheelOfDeathCounterEnchantment_TOY_529e1
		);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const controllerId = entity.GetEffectiveController();
		const turnsBeforeControllerDies = tagChange.Value;
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'WHEEL_OF_DEATH_COUNTER_UPDATED',
				GameEventHelper.CreateProvider(
					'WHEEL_OF_DEATH_COUNTER_UPDATED',
					entity.CardId,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						TurnsBeforeControllerDies: turnsBeforeControllerDies,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const showEntity = node.Object as ShowEntity;
		const controllerId = showEntity.GetEffectiveController();
		const turnsBeforeControllerDies = showEntity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_1);
		return [
			GameEventProvider.Create(
				showEntity.TimeStamp,
				'WHEEL_OF_DEATH_COUNTER_UPDATED',
				GameEventHelper.CreateProvider(
					'WHEEL_OF_DEATH_COUNTER_UPDATED',
					showEntity.CardId,
					controllerId,
					showEntity.Entity,
					this.StateFacade,
					{
						TurnsBeforeControllerDies: turnsBeforeControllerDies,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/WheelOfDeathCounterParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class WheelOfDeathCounterParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public WheelOfDeathCounterParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange;
            return stateType == StateType.PowerTaskList
                // Limit it to merceanries, the only mode where this is used, to limit the impact on the number of events sent (esp. in BG)
                && node.Type == typeof(TagChange)
                && (tagChange = node.Object as TagChange).Name == (int)GameTag.TAG_SCRIPT_DATA_NUM_1
                && GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity)?.CardId == CardIds.WheelOfDeath_WheelOfDeathCounterEnchantment_TOY_529e1;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            ShowEntity showEntity;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(ShowEntity)
                && (showEntity = node.Object as ShowEntity).CardId == CardIds.WheelOfDeath_WheelOfDeathCounterEnchantment_TOY_529e1;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var controllerId = entity.GetEffectiveController();
            var turnsBeforeControllerDies = tagChange.Value;
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "WHEEL_OF_DEATH_COUNTER_UPDATED",
                GameEvent.CreateProvider(
                    "WHEEL_OF_DEATH_COUNTER_UPDATED",
                    entity.CardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //null,
                    new {
                        TurnsBeforeControllerDies = turnsBeforeControllerDies,
                    }
                ),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var showEntity = node.Object as ShowEntity;
            var controllerId = showEntity.GetEffectiveController();
            var turnsBeforeControllerDies = showEntity.GetTag(GameTag.TAG_SCRIPT_DATA_NUM_1);
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            return new List<GameEventProvider> { GameEventProvider.Create(
                showEntity.TimeStamp,
                "WHEEL_OF_DEATH_COUNTER_UPDATED",
                GameEvent.CreateProvider(
                    "WHEEL_OF_DEATH_COUNTER_UPDATED",
                    showEntity.CardId,
                    controllerId,
                    showEntity.Entity,
                    StateFacade,
                    //null,
                    new {
                        TurnsBeforeControllerDies = turnsBeforeControllerDies,
                    }
                ),
                true,
                node) };
        }
    }
}
